File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

string solution(vector<string> participant, vector<string> completion) {
    
    // 이름, 통과한 사람 수
    unordered_map<string, int> hash_map;
    
    // 통과한 사람이 map에 있으면 해당 키에 대응하는 값을 1 증가
    // 없으면 이름을 키로, 1을 값으로 하는 새로운 엔트리 추가
    for(string c: completion) {
        auto it = hash_map.find(c);
        if (it == hash_map.end()) hash_map[c] = 1;
        else hash_map[c]++;
    }
    
    // map에 아예 없으면 바로 해당 이름 리턴
    // map에 있는데, 해당 값이 0이면 해당 이름 리턴
    // map에 있는데, 0이 아니면 값 감소 (동명이인 중 일부만 통과함)
    string answer = "";
    for(string p: participant) {
        auto it = hash_map.find(p);
        if (it == hash_map.end()) {
            answer = p;
            break;
        }
        else if (hash_map[p] == 0) {
            answer = p;
            break;
        }
        else {
            hash_map[p]--;
        }
    }
    
    return answer;
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;

class Solution {
    
    static Map<String, Integer> people = new HashMap<String, Integer>();
    
    public String solution(String[] participant, String[] completion) {
        
        // 1. 참가자들을 이름-참가자 수로 명단을 매핑한다.
        for(String p: participant) {
            if (!people.containsKey(p))
                people.put(p, 1);
            else
                people.put(p, people.get(p)+1);
        }
        
        // 2. 완주자들에 대해 명단에서 카운트를 줄인다.
        String answer = "";
        for(String c: completion) {
            int tmp = people.get(c);
            if (tmp == 1) people.remove(c);
            else people.put(c, tmp-1);
        }
        
        List<String> remains = new ArrayList<>(people.keySet());
        
        return remains.get(0);
    }
}